components/script/script_thread.rs

```diff
@@ -2071,12 +2081,33 @@ impl ScriptThread {
     fn handle_post_message_msg(
         &self,
         pipeline_id: PipelineId,
+        source_pipeline_id: PipelineId,
+        source_browsing_context: TopLevelBrowsingContextId,
         origin: Option<ImmutableOrigin>,
         data: Vec<u8>,
     ) {
         match { self.documents.borrow().find_window(pipeline_id) } {
-            None => return warn!("postMessage after pipeline {} closed.", pipeline_id),
-            Some(window) => window.post_message(origin, StructuredCloneData::Vector(data)),
+            None => return warn!("postMessage after target pipeline {} closed.", pipeline_id),
+            Some(window) => {
+                // FIXME: synchronously talks to constellation.
+                // send the required info as part of postmessage instead.
+                let source = match self.remote_window_proxy(
+                    &*window.global(),
+                    source_browsing_context,
+                    source_pipeline_id,
+                    None,
+                ) {
+                    None => {
+                        return warn!(
+                            "postMessage after source pipeline {} closed.",
+                            source_pipeline_id,
+                        )
+                    },
+                    Some(source) => source,
+                };
+                // FIXME: enqueues a task; unnecessary delay.
```

Comment:
Would it be good to write the `FIXME` as `FIXME(#22512)` to know there's an open issue for it?

Reply:
Yep!